### PR TITLE
Refactor evm

### PR DESCRIPTION
**Pull request by zjb0807**

### CODE REVIEW

**Comment by zjb0807** [comment]
Next PR will use `ethabi.encode` and `abi.decode` to improve the precompile and predeploy-contracts

**Review by xlc — COMMENTED** [review]
there are a lot of overhead of tracing so I think we want it disabled for wasm by default, and have it enabled for std

File: modules/evm/src/lib.rs [hunk 1]
```diff
@@ -791,16 +808,47 @@ pub mod module {
...
+			if account_info.contract_info.is_some() {
+				return false;
+			}
+
+			account_info.nonce.is_zero()
+		})
+	}
+
+	/// Remove an account if its empty.
+	pub fn remove_account_if_empty(address: &H160) {
```
Comment (xlc) [inline modules/evm/src/lib.rs:835]: this is not used at all?

File: modules/evm/src/lib.rs [hunk 2]
```diff
@@ -830,7 +878,7 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// Removes an account from Accounts and AccountStorages.
-	pub fn remove_account(address: &EvmAddress) -> Result<(), ExitError> {
+	pub fn remove_account(address: &EvmAddress) -> Result<(), DispatchError> {
```
Comment (xlc) [inline modules/evm/src/lib.rs:881]:
```suggestion
	pub fn remove_account(address: &EvmAddress) -> DispatchResult {
```

File: modules/evm/src/lib.rs [hunk 3]
```diff
@@ -861,75 +909,41 @@ impl<T: Config> Pallet<T> {
...
-	}
-
-	pub fn update_contract_storage_size(address: &EvmAddress, change: i32) {
-		if change == 0 {
+	/// Create an account.
+	/// - Create new account for the contract.
+	/// - Update codes info.
+	/// - Update maintainer of the contract.
+	/// - Save `code` if not saved yet.
+	pub fn create_account(source: H160, address: H160, code: Vec<u8>) {
```
Comment (xlc) [inline modules/evm/src/lib.rs:917]: should we call it `create_contract`

File: modules/evm/src/lib.rs [hunk 4]
```diff
@@ -956,9 +970,59 @@ impl<T: Config> Pallet<T> {
...
+	/// Get code hash at given address.
+	pub fn code_hash_at_address(address: &EvmAddress) -> H256 {
+		if let Some(AccountInfo {
+			contract_info: Some(contract_info),
+			..
+		}) = Self::accounts(address)
+		{
+			contract_info.code_hash
+		} else {
+			code_hash(&[])
```
Comment (xlc) [inline modules/evm/src/lib.rs:1006]: better to precompute this and hardcode it, and some unit test to test the hardcoded value

File: modules/evm/src/lib.rs (hunk 1 above)
Comment (zjb0807) [reply modules/evm/src/lib.rs:834]: Yes. Add comment.

File: modules/evm/src/lib.rs [hunk 5]
```diff
@@ -861,75 +909,41 @@ impl<T: Config> Pallet<T> {
...
-		ContractStorageSizes::<T>::mutate(address, |val| {
-			if change > 0 {
-				*val = val.saturating_add(change as u32);
-			} else {
-				*val = val.saturating_sub((-change) as u32);
-			}
+
+		// if source is account, the maintainer of the new contract is source.
+		// if source is contract, the maintainer of the new contract is the maintainer of the contract.
+		let maintainer = Self::accounts(source).map_or(source, |account_info| {
```
Comment (ermalkaleci) [inline modules/evm/src/lib.rs:926]: I'm not sure about this behaviour.  If someone uses a contract to deploy a contract then he loses the ownership.

Comment (zjb0807) [reply modules/evm/src/lib.rs:926]: Refer to this document to achieve
```
Contract Maintainer
The deployer of a contract will become the maintainer of the contract
If a contract deploy another contract, the maintainer of the newly deployed contract is the maintainer of the deployer contract
Maintainer can transfer the role to other accounts, which can be a smart contract
```

File: modules/evm/src/lib.rs (hunk 3 above)
Comment (ermalkaleci) [reply modules/evm/src/lib.rs]: maybe `create_contract_account`

File: modules/evm/src/lib.rs (hunk 5 above)
Comment (ermalkaleci) [reply modules/evm/src/lib.rs:926]: @zjb0807 I understand that, I am questioning if that's the behaviour we want

File: modules/evm/src/lib.rs [hunk 6]
```diff
@@ -791,16 +811,43 @@ pub mod module {
...
+	/// Check whether an account is empty.
+	pub fn is_account_empty(address: &H160) -> bool {
+		let account_id = T::AddressMapping::get_account_id(address);
+		let balance = T::Currency::total_balance(&account_id);
+
+		if !balance.is_zero() {
+			return false;
+		}
+
+		Accounts::<T>::mutate_exists(address, |maybe_account_info| -> bool {
```
Comment (ermalkaleci) [inline modules/evm/src/lib.rs:823]: @zjb0807 why is this `mutate_exists`?

Comment (zjb0807) [reply modules/evm/src/lib.rs]: fixed

File: modules/evm/src/lib.rs (hunk 5 above)
Comment (xlc) [reply modules/evm/src/lib.rs:922]: I don't see issue here. When a parent contract deploy a child contract, the ownership of child contract should by default belong to parent contract. It is up to parent contract to decide if it want to transfer the ownership of child contract to others.

File: modules/evm/src/lib.rs [hunk 7]
```diff
@@ -956,9 +1032,60 @@ impl<T: Config> Pallet<T> {
...
+	pub fn code_hash_at_address(address: &EvmAddress) -> H256 {
+		if let Some(AccountInfo {
+			contract_info: Some(contract_info),
+			..
+		}) = Self::accounts(address)
+		{
+			contract_info.code_hash
+		} else {
+			// The same as `code_hash(&[])`, hardcode here.
+			H256::from_slice(&hex!("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470").to_vec())
```
Comment (xlc) [inline modules/evm/src/lib.rs:1069]:
```suggestion
			H256::from_slice(&hex!("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"))
```

**Review by xlc — APPROVED** [review]

**Pull request merged**